src/Server/ClientEmbedded/ClientEmbedded.h

```diff
@@ -0,0 +1,82 @@
+#pragma once
+
+#if defined(OS_LINUX)
+
+#include <Client/ClientBase.h>
+#include <Client/LocalConnection.h>
+
+#include <Core/Settings.h>
+#include <Interpreters/Context.h>
+#include <Loggers/Loggers.h>
+#include <Common/InterruptListener.h>
+#include <Common/StatusFile.h>
+#include <Common/Config/ConfigHelper.h>
+#include <Common/Config/ConfigProcessor.h>
+
+
+#include <Poco/Util/LayeredConfiguration.h>
+#include <filesystem>
+#include <memory>
+#include <optional>
+
+
+namespace DB
+{
+
+// Client class which can be run embedded into server
+class ClientEmbedded final : public ClientBase
+{
+public:
+    explicit ClientEmbedded(
+        std::unique_ptr<Session> && session_,
+        int in_fd_,
+        int out_fd_,
+        int err_fd_,
+        std::istream & input_stream_,
+        std::ostream & output_stream_,
+        std::ostream & error_stream_)
+        : ClientBase(in_fd_, out_fd_, err_fd_, input_stream_, output_stream_, error_stream_), session(std::move(session_))
+    {
+        global_context = session->makeSessionContext();
+        configuration = ConfigHelper::createEmpty();
+        layered_configuration = new Poco::Util::LayeredConfiguration();
+        layered_configuration->addWriteable(configuration, 0);
+    }
+
+    int run(const NameToNameMap & envVars, const String & first_query);
+
+    /// NOP. The embedded client runs inside the server process which has its own signal handlers.
+    /// Thus we cannot override it in any way.
+    void setupSignalHandler() override {}
```

Comment:
@nikitamikhaylov for this ClientBase has a method  https://github.com/ClickHouse/ClickHouse/blob/e9c7f51a9ea251a7a149d1ddf0d80b54d59fed87/src/Client/ClientBase.h#L102 which is called by ClientApplicationBase on signal interrupt https://github.com/ClickHouse/ClickHouse/blob/e9c7f51a9ea251a7a149d1ddf0d80b54d59fed87/src/Client/ClientApplicationBase.cpp#L54.

But since we can't setup signal handlers in a single process with database -> we should process charecters in SSHPtyHandler and if Ctrl-C is intercepted -> stop current query by calling mentioned method in ClientBase. 

Or the other way around it is spawn client in different process and figure out a way to talk with db back and forth